src/Microsoft.AspNetCore.Mvc.Core/ModelBinding/Validation/ValidationVisitor.cs

```diff
@@ -192,6 +196,27 @@ protected virtual bool Visit(ModelMetadata metadata, string key, object model)
                 return true;
             }
 
+            if (++ValidationDepth > MaxValidationDepth)
+            {
+                // Non cyclic but too deep an object graph.
+                string displayString;
+                switch (metadata.MetadataKind)
+                {
+                    case ModelMetadataKind.Property:
+                        displayString = $"property '{metadata.Name}' on type '{metadata.ContainerType.Name}'";
+                        break;
+
+                    default:
+                        displayString = $"model of type '{metadata.ModelType.Name}'";
+                        break;
+                }
+
+                // Pop the current model to make ValidationStack.Dispose happy
+                CurrentPath.Pop(model);
+
+                throw new InvalidOperationException($"ValidationVisitor exceeded the maximum configured validation depth '{MaxValidationDepth}' when validating {displayString}. This may indicate a large or infinitely recursive object graph.");
```

Comment:
It's worth thinking about what the behavior should be here. @JamesNK - what does json.net do with it's max depth? throw?

I think if we're going to throw an exception here we also need to explain how you solve the problem by preventing recursion into certain types/properties.

Reply:
It sounds like this was discussed in the issue, and we proposed there that we throw (which I'm pretty certain is the right thing to do).

Reply:
It throws: https://www.newtonsoft.com/json/help/html/MaxDepth.htm